apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -197,6 +199,94 @@ function extractFromCompletion(
   return out;
 }
 
+/**
+ * Initializes the AI question generation cache used for rate limiting.
+ */
+let aiQuestionGenerationCache: Cache | undefined;
+export function initializeAiQuestionGenerationCache() {
+  // The cache variable is outside the function to avoid creating multiple instances of the same cache in the same process.
+  if (aiQuestionGenerationCache) return aiQuestionGenerationCache;
+  aiQuestionGenerationCache = new Cache();
+  aiQuestionGenerationCache.init({
+    type: config.nonVolatileCacheType,
+    keyPrefix: config.cacheKeyPrefix,
+    redisUrl: config.nonVolatileRedisUrl,
+  });
+  return aiQuestionGenerationCache;
+}
+
+/**
+ * Approximate the cost of the prompt, in US dollars.
+ * Accounts for the cost of prompt, system, and completion tokens.
+ */
+export function approximatePromptCost(prompt: string) {
+  // There are approximately 4 characters per token (source: https://platform.openai.com/tokenizer),
+  // so we divide the length of the prompt by 4 to approximate the number of prompt tokens.
+  // Also, on average, we generate 3750 system tokens per prompt.
+  const approxPromptAndSystemTokenCost =
+    ((prompt.length / 4 + 3750) * config.costPerMillionPromptTokens) / 1e6;
+
+  // On average, we generate 250 completion tokens per prompt.
+  const approxCompletionTokenCost = (250 * config.costPerMillionCompletionTokens) / 1e6;
+
+  return approxPromptAndSystemTokenCost + approxCompletionTokenCost;
+}
+
+/**
+ * Retrieve the Redis key for a user's current AI question generation interval usage
+ */
+function getIntervalUsageKey(userId: number) {
+  const intervalStart = Date.now() - (Date.now() % intervalLengthMs);
+  return `ai-question-generation-usage:user:${userId}:interval:${intervalStart}`;
+}
+
+// 1 hour in milliseconds
+const intervalLengthMs = 3600 * 1000;
+
+/**
+ * Retrieve the user's AI question generation usage in the last hour interval, in US dollars
+ */
+export async function getIntervalUsage({
+  aiQuestionGenerationCache,
+  userId,
+}: {
+  aiQuestionGenerationCache: Cache;
+  userId: number;
+}) {
+  return (await aiQuestionGenerationCache.get(getIntervalUsageKey(userId))) ?? 0;
+}
+
+/**
+ * Add the cost of a completion to the usage of the user for the current interval.
+ */
+export async function addCompletionCostToIntervalUsage({
+  aiQuestionGenerationCache,
+  userId,
+  promptTokens,
+  completionTokens,
+  intervalCost,
+}: {
+  aiQuestionGenerationCache: Cache;
+  userId: number;
+  promptTokens: number;
+  completionTokens: number;
+  intervalCost: number;
+}) {
+  const completionCost =
+    (config.costPerMillionPromptTokens * (promptTokens ?? 0) +
+      config.costPerMillionCompletionTokens * (completionTokens ?? 0)) /
```

Comment:
Nit, this is a little simpler:

```suggestion
    (config.costPerMillionPromptTokens * (promptTokens ?? 0 + completionTokens ?? 0) /
```

Reply:
I separated `costPerMillionPromptTokens` and `costPerMillionCompletionTokens` because prompt tokens are $3.75/million tokens and completion tokens are $15/million tokens.:

https://github.com/PrairieLearn/PrairieLearn/blob/8dd984855a6576a41270fa5e469e116866096c25/apps/prairielearn/src/lib/config.ts#L578-L582

Would you still recommend making this change, or is it fine as-is because of this?

Reply:
Ahh! No, you're exactly right, this is correct as is. Thanks for pointing that out. I missed the difference between the two variables here.

Reply:
All good 👍